mplex/src/lib.rs

```diff
@@ -18,730 +18,326 @@
-    }
-
-    #[inline]
-    fn protocol_names(&self) -> Self::NamesIter {
-        iter::once((Bytes::from("/mplex/6.7.0"), ()))
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut inner = self.inner.lock();
+        let num = try_ready!(next_match(&mut inner, |elem| {
+            match elem {
+                codec::Elem::Open { substream_id } => Some(*substream_id),       // TODO: check even/uneven?
+                _ => None,
+            }
+        }));
+
+        if let Some(num) = num {
+            Ok(Async::Ready(Some(Substream {
+                inner: self.inner.clone(),
+                current_data: Bytes::new(),
+                num,
+                endpoint: Endpoint::Listener,
+            })))
+        } else {
+            Ok(Async::Ready(None))
+        }
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use header::PacketType;
-    use std::io;
-    use tokio_io::io as tokio;
-
-    #[test]
-    fn can_use_one_stream() {
-        let message = b"Hello, world!";
-
-        let stream = io::Cursor::new(Vec::new());
-
-        let mplex = Multiplex::dial(stream);
-
-        let mut substream = mplex
-            .clone()
-            .outbound()
-            .wait()
-            .unwrap()
-            .expect("outbound substream");
-
-        assert!(tokio::write_all(&mut substream, message).wait().is_ok());
-
-        let id = substream.id();
-
-        assert_eq!(
-            substream
-                .name()
-                .and_then(|bytes| String::from_utf8(bytes.to_vec()).ok()),
-            Some(id.to_string())
-        );
-
-        let stream = io::Cursor::new(mplex.state.lock().wait().unwrap().stream.get_ref().clone());
-
-        let mplex = Multiplex::listen(stream);
-
-        let mut substream = mplex.inbound().wait().unwrap().expect("inbound substream");
-
-        assert_eq!(id, substream.id());
-        assert_eq!(
-            substream
-                .name()
-                .and_then(|bytes| String::from_utf8(bytes.to_vec()).ok()),
-            Some(id.to_string())
-        );
+/// Future to the next available outgoing substream.
+///
+/// This produces the substream only after we're finished writing the opening message.
+pub struct OutboundSubstream<C> {
+    inner: Arc<Mutex<MultiplexInner<C>>>,
+    substream_id: u32,
+}
 
-        let mut buf = vec![0; message.len()];
+impl<C> Future for OutboundSubstream<C>
+where C: AsyncRead + AsyncWrite
+{
+    type Item = Option<Substream<C>>;
+    type Error = IoError;
 
-        assert!(tokio::read(&mut substream, &mut buf).wait().is_ok());
-        assert_eq!(&buf, message);
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let elem = codec::Elem::Open { substream_id: self.substream_id };
+        let mut inner = self.inner.lock();
+        try_ready!(poll_send(&mut inner, elem));
+        let _ = inner.inner.poll_complete();        // TODO: block on this?
```

Comment:
Maybe keep a pending `Option<codec::Elem>` in `MultiplexInner` and ensure that this is sent out first. If the slot is unused and `poll_complete` is not ready, `elem` could be parked there and any future read/write will first attempt to flush it out. Cf. [yamux/src/connection.rs](https://github.com/paritytech/yamux/blob/tw/develop/src/connection.rs#L102).